// vue-styleguidist/lib/client/rsg-components/StyleguideFooter/StyleguideFooterRenderer.js
import "core-js/modules/es.string.small.js";
import React from 'react';
import PropTypes from 'prop-types';
import Markdown from 'rsg-components/Markdown';
import Styled from 'rsg-components/Styled';

var styles = function styles(_ref) {
  var fontFamily = _ref.fontFamily,
      fontSize = _ref.fontSize,
      color = _ref.color;
  return {
    root: {
      display: 'block',
      color: color.light,
      fontFamily: fontFamily.base,
      fontSize: fontSize.small
    }
  };
};

export var StyleguideFooterRenderer = function StyleguideFooterRenderer(_ref2) {
  var classes = _ref2.classes,
      homepageUrl = _ref2.homepageUrl;
  return /*#__PURE__*/React.createElement("footer", {
    className: classes.root
  }, /*#__PURE__*/React.createElement(Markdown, {
    text: "Generated with [Vue Styleguidist](".concat(homepageUrl, ")")
  }));
};
StyleguideFooterRenderer.propTypes = {
  classes: PropTypes.objectOf(PropTypes.string.isRequired).isRequired,
  homepageUrl: PropTypes.string.isRequired
};
export default Styled(styles)(StyleguideFooterRenderer);
